File: lesson_13/maps_ts/src/lesson13.ts
/**
 * Provide the solution to LeetCode 3146 here:
 * https://leetcode.com/problems/permutation-difference-between-two-strings
 */
export function findPermutationDifference(s: string, t: string): number {
    let sum = 0;
    const indexofCharS = new Map<string, number>();
    for (let i = 0; i < s.length; i++) {
        indexofCharS.set(s.charAt(i), i);
    }
    for (let i = 0; i < t.length; i++) {
        const ch = t.charAt(i);
        sum += Math.abs((indexofCharS.get(ch) ?? 0) - i);
    }
    return sum;
}

File: lesson_13/maps_java/lesson_13/maps_java/maps_app/src/main/java/com/codedifferently/lesson13/Lesson13.java
package com.codedifferently.lesson13;

import java.util.HashMap;
import java.util.Map;

/**
 * Lesson 13 solution for "Permutation Difference between Two Strings".
 */
public class Lesson13 {

    /**
     * Sum over characters of |index_in_s - index_in_t|.
     * Guard: if a char in t is not present in s, skip it.
     */
    public int findPermutationDifference(String s, String t) {
        if (s == null || t == null) return 0;

        Map<Character, Integer> pos = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            pos.put(s.charAt(i), i);
        }

        int sum = 0;
        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            Integer j = pos.get(c);
            if (j == null) {
                continue; // skip missing chars
            }
            sum += Math.abs(i - j);
        }
        return sum;
    }
}
